sdk/testing/src/main/java/io/opentelemetry/sdk/testing/assertj/SpanDataAssert.java

```diff
@@ -268,6 +272,57 @@ public SpanDataAssert hasAttributesSatisfying(Consumer<Attributes> attributes) {
     return this;
   }
 
+  /** Asserts the span has attributes matching all {@code assertions} and no more. */
+  public SpanDataAssert hasAttributesSatisfyingExactly(AttributeAssertion... assertions) {
+    return hasAttributesSatisfyingExactly(Arrays.asList(assertions));
+  }
+
+  /**
+   * Asserts the span has attributes matching all {@code assertions} and no more. Assertions can be
+   * created using methods like {@link OpenTelemetryAssertions#satisfies(AttributeKey,
+   * OpenTelemetryAssertions.LongAssertConsumer)}.
+   */
+  public SpanDataAssert hasAttributesSatisfyingExactly(Iterable<AttributeAssertion> assertions) {
+    Set<AttributeKey<?>> actualKeys = actual.getAttributes().asMap().keySet();
+    Set<AttributeKey<?>> expectedKeys =
+        StreamSupport.stream(assertions.spliterator(), false)
+            .map(AttributeAssertion::getKey)
+            .collect(Collectors.toSet());
+    assertThat(actualKeys)
+        .as("span [%s] attribute keys", actual.getName())
+        .containsExactlyInAnyOrderElementsOf(expectedKeys);
+    for (AttributeAssertion attributeAssertion : assertions) {
+      AttributeKey<?> key = attributeAssertion.getKey();
+      Object value = actual.getAttributes().get(key);
+      // Already checked all keys are present.
+      requireNonNull(value);
+      AbstractAssert<?, ?> assertion = attributeValueAssertion(key, value);
+      attributeAssertion.getAssertion().accept(assertion);
+    }
+    return this;
+  }
+
+  // The return type of these assertions must match the parameters in methods like
```

Comment:
This is the main trick. I wasn't sure I'd be able to have a single typesafe method that can accept arbitrary assertions but by having these match factories, and factories the only way to create the assertions, it seems to work. Notably, this is to avoid `Consumer<? extends AbstractAssert>`, which would not work. We're taking advantage of type erasure while ensuring safety through exposing a fixed set of factories.

Well at least I think so :) Would be good if others could reason this out too.

Reply:
I think this tracks